scipy/optimize/_pava/pava_pybind.cpp

```diff
@@ -0,0 +1,99 @@
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include <numpy/arrayobject.h>
+
+namespace py = pybind11;
+
+namespace {
+
+void pava(
+    py::array_t<double, py::array::c_style | py::array::forcecast> xa,
```

Comment:
It seems so. But then the array is not modified inplace. Not so nice.

Reply:
At any rate, this needs to be tested with various dtypes, also non C contiguous arrays.

Reply:
This is correct, the use of `py::array::forecast` causes non conforming arrays to be converted, since there is no return the incoming converted arg will not be modified.

Reply:
@amjames Could you help me out here and point me to possible solutions like
- How do I tell pybind11 to error if it's not the right dtype?
- How can I return the tuple xa, wa, ra (such that copies get returned in case a copy was triggered)?

Reply:
> * How do I tell pybind11 to error if it's not the right dtype?

The template arguments to `py::array_t` will control that behavior, particularly the flag `py::array::forecast` indicates that if the incomming array does not meet the requirements, in this case dtype must be `double` or `float64` and it must be row major and non-strided (`py::array::c_style`). You can read more about restricting or limiting these interfaces [in the docs](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#arrays).

If you drop `py::array::forecast` then pybind's overload selection mechanism will reject this function as a candidate when called with arguments without this dtype and data ordering. In the current state it will attempt to cast to the specified dtypes and copy data to contiguous memory. 

> * How can I return the tuple xa, wa, ra (such that copies get returned in case a copy was triggered)?

Pybind will automatically convert between some python and C++ types, there is a mostly complete table [here](https://pybind11.readthedocs.io/en/stable/advanced/cast/overview.html#list-of-all-builtin-conversions) but the tests for pybind are probably the most complete source of truth for that. In this case returning a `std::tuple<Txa, Twa, Tra>` will work. The placeholder template parameters would need to be replaced with the corresponding `py::array_t` instanciations. 

Type conversion will nest, so the return `std::tuple<py::array_t<...>,...>` will convert to a python `Tuple[np.ndarray, ...]`.

Reply:
Also note, that if it is desired to support other dtypes you may wish to rewrite the function as a function template and bind a few overloads. 

```C++
template<typename value_t, typename index_t>
std::tuple<
  py::array_t<value_t, py::array::c_style>, 
  py::array_t<value_t, py::array::c_style>, 
  py::array_t<index_t, py::array::c_style>,
  index_t> 
pava_worker(  py::array_t<value_t, py::array::c_style> xa, 
  py::array_t<value_t, py::array::c_style> wa, 
  py::array_t<index_t, py::array::c_style> ra ) {
...
// code here must use `index_t` and `value_t` instead of `intptr_t/size_t` and `double` 
}


m.def("pava", &pava_worker<double, int>, ...)
m.def("pava", &pava_worker<float, int>, ...)
```

The above would allow `pava` to be called with `x` and `w` either `float64` or `float32`, but not with mixed types. Dropping forecast from the `py::array_t` template arguments prevents any given overload from accepting a non-conforming array entirely though so the ordering is enforced strictly without conversion.

Reply:
@amjames Thanks for you help.
What I still do not understand is, why no error is raised when I remove the `py::array::forcecast` and pass an array with wrong dtype. (Therefore, I opened https://github.com/pybind/pybind11/discussions/4661).

Reply:
@lorentzenchr 

If you want an error to be raised you need to decorate the arguments in the binding signature with `.noconvert()` see https://pybind11.readthedocs.io/en/stable/advanced/functions.html#non-converting-arguments. 

forecast will cause an argument to be coerced before moving on to the next overload. Meaning it may try to do unsafe casts between data types. If you want the argument to check the conditions and do nothing to the argument then `py::arg("x").noconvert()` in the `m.def(...)` should do that. 

However the pybind-generated error message may not be particularly descriptive so it may be desirable to check conditions in the wrapper, or to catch the error in the wrapper and re-raise according to scipy conventions.